[feature] Add 'expand_table' feature

lavigne958 opened the pull request:
Add a new feature that allows a user to expand a cell range
into a table.

the expand will look for the right most cell with adjacent value.
the expand will look for the bottom most cell with adjacent value.
the expand will table down from top left celle range to bottom right
value.

closes #1414

Signed-off-by: Alexandre Lavigne <lavigne958@gmail.com>

alifeee commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,48 @@ def add_validation(
 
         return self.client.batch_update(self.spreadsheet_id, body)
+
+    def expand_table(
```
this function could have a simplified name. I am not sure "expand_table" is that user-friendly? perhaps just "expand"? or maybe it is good

lavigne958 replied:
I checked and in the source library where this idea comes from it's called `expand()` We could use the same name so it matches.

alifeee reviewed (changes requested):
this is a nice self-contained change. good work here :)

I think we need more tests, as I foresee a lot of edge cases, and it would be good to catch those before the feature is in gspread.

For example, what is the expectation for expanding this table from the top left?

| 1 | 1 | - | 1 |
| --- | --- | --- | --- |
| 1 | 1 | 1 | 1 |
| - | - | 1 | - |
| 1 | 1 | 1 | 1 |

This is not clear to me. Also the internal (private) `utils` expanding functions are unclear what they do.

As well, we should have a `utils` test *and* a `worksheet` test. Currently there is only `utils`

lavigne958 commented on gspread/utils.py:
```diff
@@ -979,6 +985,99 @@ def to_records(
 
 
+def _expand_right(values: List[List[str]], start: int, end: int, row: int) -> int:
+    # find rightmost value
```
true ! I was quiet short ~and lazy~ on this one :laughing: 

I updated it.

lavigne958 commented on gspread/utils.py:
```diff
@@ -979,6 +985,99 @@ def to_records(
+
+
+def _expand_bottom(values: List[List[str]], start: int, end: int, col: int) -> int:
+    # find the bottommost value
```
I updated it.

lavigne958 commented on gspread/utils.py:
```diff
@@ -979,6 +985,99 @@ def to_records(
+def _expand_bottom(values: List[List[str]], start: int, end: int, col: int) -> int:
+    # find the bottommost value
+    for rows in range(start, end):
+        print("bottom check row: {} col {}".format(rows, col))
```
true :disappointed:  :facepalm:

lavigne958 commented on gspread/utils.py:
```diff
@@ -979,6 +985,99 @@ def to_records(
+    .. note::
+
+       the ``TableDirection.table`` will first look right, then look down.
+       Any empty value in the midle of the table will be ignored.
```
true, I will rework the docstring !

here in your example: it will stop on the first empty cell on the right, so V2 is the rightmost cell. Then it will go down until D2

lavigne958 commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,48 @@ def add_validation(
+
+    def expand_table(
+        self,
+        direction: TableDirection,
```
right, that would help users, I suppose the table direction will be the most useful one.

lavigne958 commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,48 @@ def add_validation(
+    def expand_table(
+        self,
+        direction: TableDirection,
+        start_range: str = "A1",
```
I changed the name, I'm not too sure about how to call that range coordinates

lavigne958 commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,48 @@ def add_validation(
+                ['', 'B3', 'C3', '', 'E3'],
+                ['', ''  , ''  , '', 'E4'],
+            ]
+            >>> worksheet.expand_table(TableDirection.table, 'B2')
```
right I added an empty cell right in the middle.

lavigne958 commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,48 @@ def add_validation(
+        .. note::
+
+           the ``TableDirection.table`` will first look right, then look down.
+           Any empty value in the midle of the table will be ignored.
```
updated

lavigne958 commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,48 @@ def add_validation(
+        :rtype list(list): the resulting matrix
+        """
+
+        return find_table(self.get(), direction, start_range)
```
right, I'll pass the kwargs from `expand()` to `get()`. Only the arguments changing the data.

lavigne958 commented on tests/utils_test.py:
```diff
@@ -503,3 +503,42 @@ def test_to_records(self):
+
+        # table = utils.find_table(values, utils.TableDirection.table, "B2")
+        # right = utils.find_table(values, utils.TableDirection.right, "B2")
+        # down = utils.find_table(values, utils.TableDirection.down, "B2")
```
that was a leftover when I was finishing the last test, it saves me some time. 
I revert the comments

lavigne958 commented:
> this is a nice self-contained change. good work here :)
> 
> I think we need more tests, as I foresee a lot of edge cases, and it would be good to catch those before the feature is in gspread.

I thought about it and I thought:
we need to put limits to what we want to handle here:
Do we want to handle empty cells in the middle of the table ?
Do we want to handle cases when the given table is not square ? 
 
> 
> For example, what is the expectation for expanding this table from the top left?
> 
> 1	1	-	1
> 1	1	1	1
> -	-	1	-
> 1	1	1	1

I would expect something like:
- assuming the first row in the above example is data and part of the table, top left 1 is A1 coordinate:
| 1 | 1 |
| 1 | 1 |

- assuming the first row are headers we don't take them:
| 1 | 1 |


> This is not clear to me. Also the internal (private) `utils` expanding functions are unclear what they do.
> 
> As well, we should have a `utils` test _and_ a `worksheet` test. Currently there is only `utils`
so far the method in `Worksheet` class does nothing particular, but I'll add test just in case.

alifeee commented:
> I thought about it and I thought:
> we need to put limits to what we want to handle here:
> Do we want to handle empty cells in the middle of the table ?
> Do we want to handle cases when the given table is not square ?

I think a reasonable implementation is "the same as what happens when you CTRL+RIGHT and CTRL+DOWN", i.e., to stop just before empty cells.

As for whether that is what other people will expect... that's another question... if people think it should work a certain way, they may think there is a bug... which is why very clear docstrings are good

> assuming the first row are headers we don't take them:
> | 1 | 1 |

wait, huh? why are we ignoring the headers? the function should just return a `List[List]]` with all items in the table starting from the called cell?

I will look back at the code now :)

alifeee commented on gspread/utils.py:
```diff
@@ -979,6 +985,116 @@ def to_records(
+
+    Expand can be done in 3 directions defined in :class:`~gspread.utils.TableDirection`
+
+        * ``TableDirection.right``: expands only same row as starting cell on the right side up to
```
"expands right until the first empty cell"

alifeee commented on gspread/utils.py:
```diff
@@ -979,6 +985,116 @@ def to_records(
+
+        * ``TableDirection.right``: expands only same row as starting cell on the right side up to
+          first null/empty cell
+        * ``TableDirection.down``: expands only same column as starting cell to the bottom up to
```
"expands down until the first empty cell"

alifeee commented on gspread/utils.py:
```diff
@@ -979,6 +985,116 @@ def to_records(
+          first null/empty cell
+        * ``TableDirection.down``: expands only same column as starting cell to the bottom up to
+          first null/empty cell
+        * ``TableDirection.table``: expands in both direction, first right then down.
```
"expands right until the first empty cell, then down until the first empty cell"

alifeee commented on gspread/utils.py:
```diff
@@ -979,6 +985,116 @@ def to_records(
+        rightMost = _expand_right(values, col, len(values[row]), row)
+
+        checkColumn = rightMost
+        if checkColumn != 0:
```
is it possible for checkColumn to be 0?

alifeee commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,57 @@ def add_validation(
+        top_left_range_name: str = "A1",
+        direction: TableDirection = TableDirection.table,
+        value_render_option: Optional[ValueRenderOption] = None,
+        date_time_render_option: Optional[DateTimeOption] = None,
```
I think maybe we do not want to duplicate all the options for `get`. We would rather provide the user no choice and say

"if you want to get different values, use `get` as you desire, and then use `utils.expand_table`"

as there are already missing options (e.g., `combine_merged_cells`, `pad_values`), and more may be added in future

so we remove value_render_option and date_time_render_option in favour of telling user to use `get` themself. what do you think?

alifeee commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,57 @@ def add_validation(
+              first null/empty cell
+            * ``TableDirection.down``: expands only same column as starting cell to the bottom up to
+              first null/empty cell
+            * ``TableDirection.table``: expands in both direction, first right then down.
```
see comments on `utils` docstring

alifeee commented on gspread/worksheet.py:
```diff
@@ -3336,3 +3338,57 @@ def add_validation(
+
+        values = self.get(
+            value_render_option=value_render_option,
+            date_time_render_option=date_time_render_option,
```
what happens if `get` does not return a 2D array?

for example if it returns

<pre>
- - - d
- a
- t
- - - a
</pre>

then will functions break? should we provide `pad_values=True`?

alifeee reviewed (commented):
you are doing good work on this change :)

I have left some comments. I am a little worried that a new feature will result in something we do not consider, causing errors. But I think we will be ok now.

thanks!!!

alifeee commented:
I have added some new tests. In particular, they are "what should happen" tests. They are:

### 1 - missing header item

```python
        values = [
            ["A1", "", "C1", ""],
            ["A2", "B2", "C2", ""],
            ["A3", "B3", "C3", ""],
            ["", "", "", ""],
        ]
        expected_table = [
            ["A1"],
            ["A2"],
            ["A3"],
        ]
```

### 2 - missing initial cell

```python
        values = [
            ["", "B1", "C1", ""],
            ["A2", "B2", "C2", ""],
            ["A3", "B3", "C3", ""],
            ["", "", "", ""],
        ]
        expected_table = [
            ["", "B1", "C1"],
            ["A2", "B2", "C2"],
            ["A3", "B3", "C3"],
        ]
```

### 3 - missing first column item

```python
        values = [
            ["A1", "B1", "C1", ""],
            ["", "B2", "C2", ""],
            ["A3", "B3", "C3", ""],
            ["", "", "", ""],
        ]
        expected_table = [
            ["A1", "B1", "C1"],
            ["", "B2", "C2"],
            ["A3", "B3", "C3"],
        ]
```

### 4 -  missing last column item

```python
        values = [
            ["A1", "B1", "C1", ""],
            ["A2", "B2", "", ""],
            ["A3", "B3", "C3", ""],
            ["", "", "", ""],
        ]
        expected_table = [
            ["A1", "B1", "C1"],
        ]
```

## Conc.

Are these what you expect? Personally I might want case 4 to end up in the entire table, as I reckon I am more likely to have a ***full*** first column, and empty gaps in the final column.

However, if I use the `CTRL+RIGHT` and `CTRL+DOWN` shortcuts in most spreadsheet software, then the tables above are the same. So I am happy with the behaviour above.

I am ready to merge this 👍. However, we should also add it to the list of examples and documentation somewhere.

alifeee reviewed (changes requested):
include in examples docs. otherwise approve :)

lavigne958 commented:
thank you for the extra tests ! :star: 

I am a bit confused with the tests and what I see when I try it online using keyboard controls :disappointed: 

1. in this case, I agree with you, when I run the same data set with `CTRL+RIGHT` the cell jumps to `C1`... which is not expected at all.... I believe here we should keep our behavior it makes more sense to me according to our documentation and what I would expect from the function
2. I disagree, the first cell is empty, then the array stops here, and returns an empty array :roll_eyes:  I checked online with keyboard controls and for some reasons it jumps to celle `B1` :facepalm: then when you go down it goes to `B3` (as expected)... I don't know what choice is the best here :disappointed: 
3. I completely agree with it, as expected, like the documentation says: we can't scan for empty cells in the middle of the table.
4. I completely agree with it, though online when I try, it get to `C1`, nice, then I use `CTRL+DOWN` then it jumps to `C3` :facepalm: I believe our example is more logic and provides the expected result. 

so for 1., 3. and 4. I agree with your tests and the result is what I would expect.
what is left is case 2., do you think if we start on a blank cell we should start looking right/down for the next cell if it has a value or look right/down for the first non empty cell and start from here ? :thinking:

alifeee commented:
I have made this sheet

https://docs.google.com/spreadsheets/d/1huMwgagFCGTMVBf5mONk_W74-qHZUnpASHuVHnU6Hsk/edit?usp=sharing

![image](https://github.com/burnash/gspread/assets/13833017/4675d614-95cf-43fd-992f-064d20933f4d)

I think maybe we do not "do the same as CTRL+arrow keys does", as this is confusing. I think we find another rule to follow (perhaps "the same as what xlwings does")

lavigne958 commented:
Hey thanks for the working sample ! I completely agree with you some behaviors are completely unexpected :disappointed: 

To me we should go for what feels logical to us, which seems to be what is described in the `xlwing` library:
- going right until first empty cell is found
- going then down until first empty cell is found.
- ignore empty starting cell (like you mentioned)
- table direction should be: _down then right_ as mention in their documentation here: [expand parameters](https://docs.xlwings.org/en/latest/api/range.html#id14)

does that sounds good to you ? :thinking: 
after all we can chose the implementation we like as well, regardless of what the Spreadsheet UI in a web browser does :laughing:

lavigne958 commented on gspread/utils.py:
```diff
@@ -979,6 +985,130 @@ def to_records(
+        if rows >= len(values):
+            return len(values) - 1
+
+        # this row is smaller than the others, just keep looking
+        if col >= len(values[rows]):
+            continue
+
+        if values[rows][col] == "":
+            return rows
```
here I noticed we have rows smaller than the starting row sometimes.
I assumed we should keep looking for rows bellow, but I believe this is wrong.

we should look down for any empty value `""` OR any row smaller than starting row right ?

alifeee replied:
I prefer to answer questions by writing tests. not sure exactly what you mean

lavigne958 replied:
what I mean is currently we can have a response from the API that looks like this:

||A |B|C|
|-|-| - | -|
|1|v1|v2|v3|
|2|v4|v5| |
|3|v6|v7|v8|

and currently in the code we:
1. start looking right from A1, until C1 (latest value in the row)
2. then we start to look down.
3. when we check C2, the row 2 returned by the API is smaller than the starting row 1
4. currently we just ignore it and keep looking to the row below
5. I believe that's wrong, we should stop. 

So the code `if col >= len(values[rows]):` is wrong we should not continue but instead consider it as an empty cell.

alifeee replied:
yes. With my suggestion (look for full top row & left column), this should not matter(?)

alifeee commented:
I believe my preference is to "find a table" by "complete row 1 and column 1"

so, to look at these `h` columns

| h | h | h | h |
| --- | --- |  ---  | --- |
| h | - | - | - |
| h | - | - | - |
| h | - | - | - |

...and if they are empty, terminate the table there.

I have updated the tests to do this (please have a look at the tests to see if you agree). What do you think?

lavigne958 commented:
> I believe my preference is to "find a table" by "complete row 1 and column 1"
> 
> so, to look at these `h` columns
> 
> h	h	h	h
> h	-	-	-
> h	-	-	-
> h	-	-	-
> ...and if they are empty, terminate the table there.
> 
> I have updated the tests to do this (please have a look at the tests to see if you agree). What do you think?

I agree with this, the only detail would be: look down first then right. 
to follow what xlwings does. it's completely arbitrary but it makes us ISO with the xlwings behavior.

lavigne958 commented:
I did quite a big rework on the iteration functions.

all tests now pass, though the tests already here needed to be adjusted: 
- we now look down first
- then right
- so the expected result changes.

it works as expected, like xlwings.

alifeee commented:
why does xlwings use the left column and *bottom* row?

it makes more sense to me that one would desire the *top* row?

lavigne958 commented:
I know, that's what I did first, but we agreed on making it ISO with xlwings.

I don't mind changing it, I just want us to agree on one implementation.

alifeee commented:
I don't think that's how xlwings works. I think xlwings uses the top row and left column.

See here for their implementation: https://github.com/xlwings/xlwings/blob/a159cc9cb7ef5f168ad0d902934404bce4a76386/xlwings/expansion.py#L31-L52

(I can't parse it with my mind fully but it looks to use the top row, not the bottom)

see also this comment: https://github.com/xlwings/xlwings/issues/557#issuecomment-242794085

lavigne958 commented:
> I don't think that's how xlwings works. I think xlwings uses the top row and left column.
> 
> See here for their implementation: https://github.com/xlwings/xlwings/blob/a159cc9cb7ef5f168ad0d902934404bce4a76386/xlwings/expansion.py#L31-L52
> 
> (I can't parse it with my mind fully but it looks to use the top row, not the bottom)
> 
> see also this comment: [xlwings/xlwings#557 (comment)](https://github.com/xlwings/xlwings/issues/557#issuecomment-242794085)

you're right ! 
same I can't parse the code with my mind, but I understand the same thing and it makes sense. to start from the given cell, go right until empty cell, from starting cell again, go down until empty cell.

this provides the right and down boundaries then the rest is just to extract the matrix from here.

I'll update the code, expect potentially a big code change/refactoring then !

lavigne958 commented:
## Update

- reworked the way we look for table, look right from starting cell, then look down **from** starting cell
- updated the tests suite to reflect the behavior
  - empty starting cell will return an empty list `[]`
  - empty cell on the way right or down will stop the search
  - empty cell in the **middle** of the matrix will be ignored
  - in the assert functions in the test, iterate over the expected result and expect the values to match
    - I notice tests passing when it should not because we iterate over the result and it's empty so we don't compare anything
- updated the doc a bit
- added extra checks on the size of the matrix and the starting cell coordinates in order to raise Exception if starting cell is outside the range of the given matrix.

I hope this covers any wrong inputs, the expected behavior from all (most?) users.

alifeee commented:
this looks all good to me!

only final thing is that I would expect a table like

| | 1 | 2 | 3 |
| --- | --- | --- | --- |
| a | 0.342 | 0.145 | 0.166 |
| b | 0.982 | 0.576 | 0.600 |
| c | 0.340 | 0.285 | 0.585 |

to expand properly, whereas it seems you suggest with

> empty starting cell will return an empty list []

that it would return as an empty list

lavigne958 commented:
Yes to me this function will return the table starting from the given coordinates. The starting cell is part of the resulting table. so if the starting cell is indeed empty then the table will be empty.

So far that seems to be the last key point where we can't find seems to be best. Should we just put it as an option like "ignore empty starting cell" ? Of true then we follow your choice of false we follow mine. This way the user decides the behavior.

What do you think?

alifeee commented:
apologies for the delay

while I think a special case for a blank row 0 column 0 would be nice (e.g., for a labelled adjacency matrix), it is probably not worth the complication (e.g., if the row 0 column 0 is empty, but also row 1 column 0 and row 0 column 1 are empty, it should return `[]` not `[[""]]`)

Thus, I think we return `[]` as you say. If people want their table to be parsed, they should fill the first cell with e.g., `"-"`, or only call the function on the table contents, and use `get_row` and `get_column` (e.g.,) to get the headings.

However, there should be a test to validate this behaviour.

After that, I think this is ready to merge!

lavigne958 commented:
I added a new commit with a test to validate the top left cell empty scenario.
In case we find a top left empty cell we return an empty list `[]`.
Is this the test you referred to in you example ? if not let me know I'll add extra tests if need be.

I added an extra test too in the basic test for find_table to check when we reach the end of the list and the end of the matrix. just in case.

alifeee commented:
this is good with me!

with the amount of work spent on a new feature... let's put reference to it in all the relevant places in the documentation :)

lavigne958 commented:
> this is good with me!
> 
> with the amount of work spent on a new feature... let's put reference to it in all the relevant places in the documentation :)

great call, it's done :heavy_check_mark: 
I tried to make it minimal with all the important details.

lavigne958 commented:
before merging I rebased the whole thing in order to solve conflicts of any.

alifeee reviewed (approved):
seems good to me ! I am assuming you have manually tested the worksheet function (as there is only a utils function)

I hope you have had fun with this one ;]

I say we can merge this.

lavigne958 commented:
yes I did :wink: it works like a charm. I just pushed a quick fix on the docstring to make the link to the raised exception work in the documentation.
